### ISSUE TITLE
Shoule feature budget calculated before or after feature normalization?

### ISSUE THREAD

**Issue by Gmrider13** [jumxglhf/G2A2C#2]
Hi, thank you for your excellent work! I have some confusion regarding the calculation of `env.feature_budget`:
In line 67 of attack_utils.py:
`self.feature_budget = (self.g.ndata['feat']>0).float().sum(1).mean() if self.discrete_feat else self.g.ndata['feat'].sum(1).mean()`
Note in line 39, the feature is normalized:
`self.g.ndata['feat'] = F.normalize(self.g.ndata['feat'])`

In the dataset with discrete features ,the `self.feature_budget` is calculated through the original feature while in continuous features, the budget is calcluated through normalized feature.

Why is there a difference in the calculation of `env.feature_budget`?

**Comment by jumxglhf** [comment]
Hi! 
Thanks for your interest in our project. This is a good point. We normalize node features for graphs as a pre-processing technique broadly explored for node classification.  It generally improves the classification performance of GNNs. I just tried some quick experiments with this line (i.e., line 39 in attack_utils.py) commented out and the conclusions are still similar and consistent.

**Comment by Gmrider13** [comment]
Thanks for replying! I also noticed that the fake node is normalized in line 34 of model.py, and the feature loss is computed through normalized fake nodes and unnormalized `feature_budget`:
in line 34 of model.py:
` feat = F.normalize(feat, dim=0)
  return feat, feat.mean(), log_prob`
in line 157 of actor_critic.py:
`feature_loss += F.mse_loss(num_feat, torch.floor(env.feature_budget*args.feature_budget)/env.feature_dim)/node_budget                
`
When both normalizations are commented out, the result is:
Epoch: 1, Current Accuracy: 22.20 (222/1000), Episode Reward: 96.05, feature loss 0.3379072844982147: 100%|██████████| 803/803 [00:19<00:00, 42.17it/s]
Epoch: 2, Current Accuracy: 23.20 (232/1000), Episode Reward: 123.67, feature loss 0.5722240805625916: 100%|██████████| 803/803 [00:19<00:00, 41.87it/s]
Epoch: 3, Current Accuracy: 23.40 (234/1000), Episode Reward: 142.37, feature loss 0.5440735220909119: 100%|██████████| 803/803 [00:19<00:00, 41.14it/s]
Epoch: 4, Current Accuracy: 23.30 (233/1000), Episode Reward: 162.95, feature loss 0.6473902463912964: 100%|██████████| 803/803 [00:20<00:00, 40.11it/s]
Early Stopping. Current Accuracy is 23.20.
Solution saved to 
Dataset: cora Victim Model: gcn, Average ACC: 23.34, STD: 0.28

The feature loss is large and difficult in descending. How to get the method to work correctly?

**Comment by jumxglhf** [comment]
Thanks for your thorough results! Since the method was developed w.r.t. normalized features, I could not think of a good solution for "getting the method to work correctly" for now. A potential reason why this is happening is that because features are not normalized, the network might struggle to generate adversarial features with a big column-wise variance. 

As far as I can think, a ghetto way of getting the feature loss to descend could be manually increasing the weight of the feature loss so it gets a larger gradient descent. But this could also affect the misclassification rate, as a trade-off between the effectiveness and budgets. 

I will accordingly update the arxiv paper with the normalization pre-processing I have used in our implementation and thanks for your contribution!

**Comment by Gmrider13** [comment]
I will try different weights, thanks for your kind response!
